### ISSUE TITLE
PHP7: Parser errors in PHPUnit do not show stacktrace

### ISSUE THREAD

**Issue by raul338** [cakephp/cakephp#9192]
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.12
* Platform and Target:
```
PHP 7.0.9 (cli) (built: Jul 27 2016 19:18:12) ( NTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
    with Xdebug v2.4.0, Copyright (c) 2002-2016, by Derick Rethans

PHPUnit 5.4.8 by Sebastian Bergmann and contributors.
```

### What you did
Install cakephp, in PagesControllerTest.php deleted `}` in the end of the file
Run PHPUnit

### Expected Behavior
Like in php 5.6.24 (ubuntu 14.04-lts)
```
PHP Parse error:  syntax error, unexpected end of file, expecting function (T_FUNCTION) in /home/raul/test/tests/TestCase/Controller/PagesControllerTest.php on line 74
PHP Stack trace:
PHP   1. {main}() /home/raul/.composer/vendor/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() /home/raul/.composer/vendor/phpunit/phpunit/phpunit:47
PHP   3. PHPUnit_TextUI_Command->run() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:113
PHP   4. PHPUnit_TextUI_Command->handleArguments() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:124
PHP   5. PHPUnit_Util_Configuration->getTestSuiteConfiguration() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:745
PHP   6. PHPUnit_Util_Configuration->getTestSuite() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Configuration.php:852
PHP   7. PHPUnit_Framework_TestSuite->addTestFiles() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Configuration.php:941
PHP   8. PHPUnit_Framework_TestSuite->addTestFile() /home/raul/.composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:413
PHP   9. PHPUnit_Util_Fileloader::checkAndLoad() /home/raul/.composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:339
PHP  10. PHPUnit_Util_Fileloader::load() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Fileloader.php:38

Parse error: syntax error, unexpected end of file, expecting function (T_FUNCTION) in /home/raul/test/tests/TestCase/Controller/PagesControllerTest.php on line 74

Call Stack:
    0.0001     231784   1. {main}() /home/raul/.composer/vendor/phpunit/phpunit/phpunit:0                                                                                                    
    0.0064    1031424   2. PHPUnit_TextUI_Command::main() /home/raul/.composer/vendor/phpunit/phpunit/phpunit:47                                                                             
    0.0064    1032048   3. PHPUnit_TextUI_Command->run() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:113                                                              
    0.0064    1035200   4. PHPUnit_TextUI_Command->handleArguments() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:124                                                  
    0.0335    5683936   5. PHPUnit_Util_Configuration->getTestSuiteConfiguration() /home/raul/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:745                                    
    0.0336    5685200   6. PHPUnit_Util_Configuration->getTestSuite() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Configuration.php:852                                             
    0.0359    5897080   7. PHPUnit_Framework_TestSuite->addTestFiles() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Configuration.php:941                                            
    0.0359    5897944   8. PHPUnit_Framework_TestSuite->addTestFile() /home/raul/.composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:413                                            
    0.0359    5938568   9. PHPUnit_Util_Fileloader::checkAndLoad() /home/raul/.composer/vendor/phpunit/phpunit/src/Framework/TestSuite.php:339                                               
    0.0360    5938704  10. PHPUnit_Util_Fileloader::load() /home/raul/.composer/vendor/phpunit/phpunit/src/Util/Fileloader.php:38
```

### Actual Behavior
```
Exception: (ParseError) - syntax error, unexpected end of file, expecting function (T_FUNCTION) in [/home/raul/test/testapp/vendor/cakephp/cakephp/src/Error/BaseErrorHandler.php, line 161]
```
I don't know if this is a cake specific bug, nor a phpunit misconfiguration on php7. But the PHP7ErrorException wrapping in unit testing hides the origin of the error and the stacktrace is not printed.

In reality i found the error when a test throw a "Invalid numeric literal" parser error, there was a `09` in a fixture/seed.

**Comment by lorenzo** [comment]
This happens when you execute a controller test? (integrationTestCase)

**Comment by thinkingmedia** [comment]
This is correct behavior.

In PHP7 fatal errors are now throwable, but the CakePHP framework only handles thrown objects that are derived from `Exception`. This is for backward compatibility with PHP5/6.

The CakePHP framework catches the `Error` and wraps as a new `Exception` but there can be no stack trace since `Exception` does not accept an inner exception of type `Error`.

I would argue that this is wrong, and breaks backward behavior with PHP5/6.

Fatal errors in PHP5/6 are not catchable, but in PHP7 they are if you handle the `Error` type.

CakePHP does not handle the `Error` type and catching then wrapping that fatal error as an `Exception` is fundamentally altering the behavior of how fatal errors are handled in Cake. The behavior should not change from PHP5/6.

Instead of wrapping an `Error` into an `Exception` the error handler should just let the fatal error happen normally (i.e. do not handle it).

This behavior would allow unit tests to execute the same in PHP5/6/7

We need to have a more indepth discussion about what proper support for PHP7 means in CakePHP, and we might want to wait until some issues with PHP7 are fixed (i.e. how they handle __toString fatal errors).

**Comment by raul338** [comment]
@lorenzo yes, the example given was a IntegrationTestCase
I also test this code
```
<?php
namespace App\Test\TestCase\Controller;

use PHPUnit\Framework\TestCase;

class SampleTest extends TestCase
{
    public function testPushAndPop()
    {
        $this->assertEquals(09, 9);
    }
}
```
Output in cakephp project
```
$ phpunit tests/TestCase/Controller/SampleTest.php 
Exception: (ParseError) - Invalid numeric literal in [/home/raul/test/testapp/vendor/cakephp/cakephp/src/Error/BaseErrorHandler.php, line 161]
```
Output just the isolated test (without a cake app, also the namespace line was removed)
```
$ phpunit SampleTest.php 
PHP Parse error:  Invalid numeric literal in /home/raul/test/SampleTest.php on line 9
```

I don't know if there are other fatal errors that might need a stacktrace (a parse error is easy to fix) so maybe in bootstrap for unit testing these fatal errors should not be catch or be more verbose in transform the Error into an Exception

**Comment by markstory** [comment]
@thinkingmedia It is hard to argue that this is 'correct' as it is quite confusing. It sounds like you know which code needs to change. Would you be willing to open a pull request?

**Comment by thinkingmedia** [comment]
@markstory I don't know if I can fix this.

My first thought was to re-throw any fatal errors which would only happen in PHP 7, but after googling I found that you can not throw exceptions from the `set_exception_handler` as it will generate a stack trace error.

So in PHP 7 we are stuck receiving these *new types* of errors in the handler. Which didn't happen before in PHP 5/6.

I can only think of doing one of the following.

1. Don't change anything. 
2. Remove all type hinting for `Exception` in the Cake core so that the `set_exception_handler` can work on PHP 7 with `Error` classes.
3. Disable the exception handler when running unit tests and let PHPUnit handle it.

**Comment by markstory** [comment]
In test cases, I don't think `set_exception_handler` is invoked/used. The [IntegrationTestCase](https://github.com/cakephp/cakephp/blob/master/src/TestSuite/IntegrationTestCase.php#L373-L376) is catching all exceptions and invoking the configured error handlers.

**Comment by thinkingmedia** [comment]
The error message

```
Exception: (ParseError) - syntax error, unexpected end of file, expecting function (T_FUNCTION) in [/home/raul/test/testapp/vendor/cakephp/cakephp/src/Error/BaseErrorHandler.php, line 161]
```

Appears to be formatted the same as the `PHP7ErrorException` which is only used by the `set_exception_handler`.

See the constructor

```
    public function __construct($error)
    {
        $this->_error = $error;
        $message = $error->getMessage();
        $code = $error->getCode();
        parent::__construct(sprintf('(%s) - %s', get_class($error), $message), $code);
    }
```

I'm making an assumption that `'(%s) - %s'` wasn't used by another exception as I couldn't find it.

That class is thrown by `BaseErrorHandler::wrapAndHandleException()` and no where else. So it looks like it is being caught during unit tests.

I think the best way to fix this is to remove the `Exception` type hinting from the `BaseErrorHandler` class and `ConsoleErrorHandler`, then remove the `PHP7ErrorException` class.

This changes the method signature of `BaseErrorHandler` and people will need to modify their code if they are using it. 

I'm just afraid that making such a change would break other things.

If you read the "expected behavior" of this issue. It's showing the fatal error trace from PHP 5.6, but in PHP 7 these fatal errors are now caught by the error handler.

This is a very common issue for PHP 7. I'm finding countless open source projects that are breaking because of this change. I think this fix would have to go into the `3.next` branch.

**Comment by markstory** [comment]
I'll take a look at this, I have PHP7 installed locally. And while I'd rather not break compatibility, we can do better than what we currently have.

**Comment by thinkingmedia** [comment]
I got stack traces today running my unit tests. So the above must be related to testing controllers. Which means that Cake's error handler is loaded and if we change this we impact fatal errors outside of PHPUnit.

So this issue is more about how Cake handles fatal errors in it's error handle than anything with PHPUnit.

```
$ vendor/bin/phpunit
PHPUnit 5.4.8 by Sebastian Bergmann and contributors.

.......................III..II.................................  63 / 125 ( 50%)
.PHP Fatal error:  Class App\Model\Entity\DataItem contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (App\Model\IDataSourceChild::getUrl) in C:\work\tags\web\src\Model\Entity\DataItem.php on line 32
PHP Stack trace:
PHP   1. {main}() C:\work\tags\web\vendor\phpunit\phpunit\phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() C:\work\tags\web\vendor\phpunit\phpunit\phpunit:47
PHP   3. PHPUnit_TextUI_Command->run() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\Command.php:113
PHP   4. PHPUnit_TextUI_TestRunner->doRun() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\Command.php:162
PHP   5. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\TestRunner.php:465
PHP   6. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
PHP   7. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
PHP   8. PHPUnit_Framework_TestCase->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
PHP   9. PHPUnit_Framework_TestResult->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:892
PHP  10. PHPUnit_Framework_TestCase->runBare() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestResult.php:701
PHP  11. PHPUnit_Framework_TestCase->runTest() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:936
PHP  12. ReflectionMethod->invokeArgs() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:1085
PHP  13. Gems\Test\TestCase\Rules\OlderThanDaysTest->testAccept() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:1085
PHP  14. spl_autoload_call() C:\work\tags\web\plugins\Gems\tests\TestCase\Rules\OlderThanDaysTest.php:17
PHP  15. Composer\Autoload\ClassLoader->loadClass() C:\work\tags\web\plugins\Gems\tests\TestCase\Rules\OlderThanDaysTest.php:17
PHP  16. Composer\Autoload\includeFile() C:\work\tags\web\vendor\composer\ClassLoader.php:301
PHP  17. include() C:\work\tags\web\vendor\composer\ClassLoader.php:412

Fatal error: Class App\Model\Entity\DataItem contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (App\Model\IDataSourceChild::getUrl) in C:\work\tags\web\src\Model\Entity\DataItem.php on line 32

Call Stack:
    0.0003     365744   1. {main}() C:\work\tags\web\vendor\phpunit\phpunit\phpunit:0
    0.2888    1596856   2. PHPUnit_TextUI_Command::main() C:\work\tags\web\vendor\phpunit\phpunit\phpunit:47
    0.2889    1600016   3. PHPUnit_TextUI_Command->run() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\Command.php:113
    1.9735    8477416   4. PHPUnit_TextUI_TestRunner->doRun() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\Command.php:162
    2.0483    8752496   5. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\TextUI\TestRunner.php:465
    2.0701    8759840   6. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
    4.5894   15500928   7. PHPUnit_Framework_TestSuite->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
    4.5939   15501552   8. PHPUnit_Framework_TestCase->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestSuite.php:753
    4.5940   15501552   9. PHPUnit_Framework_TestResult->run() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:892
    4.5940   15501552  10. PHPUnit_Framework_TestCase->runBare() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestResult.php:701
    4.5946   15538800  11. PHPUnit_Framework_TestCase->runTest() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:936
    4.5946   15539096  12. ReflectionMethod->invokeArgs() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:1085
    4.5946   15539104  13. Gems\Test\TestCase\Rules\OlderThanDaysTest->testAccept() C:\work\tags\web\vendor\phpunit\phpunit\src\Framework\TestCase.php:1085
    4.6020   15543496  14. spl_autoload_call() C:\work\tags\web\plugins\Gems\tests\TestCase\Rules\OlderThanDaysTest.php:17
    4.6020   15543632  15. Composer\Autoload\ClassLoader->loadClass() C:\work\tags\web\plugins\Gems\tests\TestCase\Rules\OlderThanDaysTest.php:17
    4.6025   15543728  16. Composer\Autoload\includeFile() C:\work\tags\web\vendor\composer\ClassLoader.php:301
    4.6081   15547008  17. include('C:\work\tags\web\src\Model\Entity\DataItem.php') C:\work\tags\web\vendor\composer\ClassLoader.php:412
```

@markstory I think I might take a crack a PR later today unless you've got something you want to try.

**Comment by markstory** [comment]
@thinkingmedia I've not done any work on this yet. I was going to take a look tonight time providing.

**Comment by markstory** [comment]
Fixed with #9209
